### PR TITLE
Server: extension fields

### PR DESCRIPTION
**Pull request by georg-schwarz**
Addresses backend part of #18.

It is unclear to me yet, how and if we can include that into the generated API client library.

### CODE REVIEW

**Comment by MBuchalik** [comment]
Disclaimer: I don't have much experience with Nestjs, Swagger or the client generator. 

I found no way to do this with class-validator. Even though nestjs/swagger supports this: https://github.com/nestjs/swagger/issues/1170

One alternative I saw would be to create a separate field like "customData", and to define a type for it as described here: https://stackoverflow.com/questions/68408494/how-to-make-key-string-string-appear-in-nestjs-swagger-generation (A similar approach: https://github.com/typestack/class-validator/issues/795)

In our case, I guess we have to use something like this (not tested!):

```
@ApiProperty({
  type: 'object',
  additionalProperties: {
    oneOf: [
      { type: 'string' },
      { type: 'number' },
      { type: 'boolean' },
      {
        type: 'array',
        items: {
          oneOf: [
            { type: 'string' },
            { type: 'number' },
            { type: 'boolean' },
          ],
        },
      },
    ],
  },
})
customData: Record<string, string | number | boolean | Array<string | number | boolean>>;
```

Originally, I wanted to use something like this, but it seems like the TS types of `@ApiProperty` do not allow this:

```
@ApiProperty({
  type: 'object',
  additionalProperties: {
    type: [
      'string',
      'number',
      'boolean',
      { type: 'array', items: { type: ['string', 'number', 'boolean'] } },
    ],
  },
})
customData: Record<string, string | number | boolean | Array<string | number | boolean>>;
```

**Comment by georg-schwarz** [comment]
Yeah, I tried a few things to get it to work...

> One alternative I saw would be to create a separate field like "customData" [...]
Also not that easy to implement with class-validator.

I think we should go with the current approach and refactor the client library. Maybe rewrite without code generation but ensure correctness with tests or check messages with zod.

**Comment by georg-schwarz** [comment]
I think I did it! :)

File: server/src/service-docs/extensions.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  const extensionKeys = Object.keys(extensionObject);
+  for (const extensionKey of extensionKeys) {
+    const extensionValue = (extensionObject as Record<string, unknown>)[
+      extensionKey
+    ];
+    if (!isExtensionValueType(extensionValue)) {
+      return false;
+    }
+  }
+  return true;
```
Comment (MBuchalik) [inline server/src/service-docs/extensions.ts:52]:
```suggestion
  for (const extensionValue of Object.values(extensionObject)) {
    if (!isExtensionValueType(extensionValue)) {
      return false;
    }
  }
  return true;
```

File: server/src/service-docs/service-docs.service.spec.ts [hunk 2]
```diff
@@ -102,3 +109,62 @@ describe('ServiceDocsService', () => {
...
+
+describe('ServiceDocs ORM conversion', () => {
+  it('should convert valid extension fields', () => {
+    const model = {
+      name: 'test',
+      extensions: {
+        'x-test1': 123,
+        'x-test2': 'asd',
+        'x-test3': false,
+        'x-test4': [123, 'asd', false],
```
Comment (MBuchalik) [inline server/src/service-docs/service-docs.service.spec.ts:121]: I think we don't need the "x" prefix anymore. Should we maybe remove this prefix from the tests as well?

The same goes for the e2e test.

File: server/src/service-docs/extensions.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    for (const item of extensionValue) {
+      if (!isExtensionPrimitiveValueType(item)) {
+        return false;
+      }
+    }
+  } else if (!isExtensionPrimitiveValueType(extensionValue)) {
+    return false;
+  }
+
+  return true;
```
Comment (MBuchalik) [inline server/src/service-docs/extensions.ts:33]: This is just an optional code style improvement - if you like your version better, then just ignore this :D 

```suggestion
  if (Array.isArray(extensionValue)) {
    for (const item of extensionValue) {
      if (!isExtensionPrimitiveValueType(item)) {
        return false;
      }
    }

    return true;
  }

  return isExtensionPrimitiveValueType(extensionValue);
```

**Review by MBuchalik — COMMENTED** [review]
Looks good! 🎉 Happy to hear that the client generation finally works.

I added 3 comments, but they are just minor nitpicks. Oh, and I noticed that the linter is failing.

**Comment by MBuchalik** [comment]
Oh, and I noticed that in our root README it says 
```
Extend the documentation format to your needs (Upcoming).
```

I think we can remove the "Upcoming", right? (Even though the frontend is still missing.)

**Pull request merged**